File: python/clock/clock.py


class Clock():
    def __init__(self, hours, minutes):
        self.minute_of_the_day = (hours * 60 + minutes) % (24 * 60)

    def add(self, minutes):
        return Clock(0, self.minute_of_the_day + minutes)

    def _hours_and_minutes(self):
        return (self.minute_of_the_day // 60, self.minute_of_the_day % 60)

    def __hash__(self):
        return hash(self.minute_of_the_day)

    def __eq__(self, other):
        return self.minute_of_the_day == other.minute_of_the_day

    def __str__(self):
        return '%02d:%02d' % self._hours_and_minutes()

File: go/clock/clock.go
package clock

import "fmt"

// The value of testVersion here must match `targetTestVersion` in the file
// clock_test.go.
const testVersion = 4

const minutesInDay = 24 * 60

type Clock struct {
	minute_of_the_day int
}

func New(hour, minute int) Clock {
	var minutes = (hour * 60 + minute) % minutesInDay
	if minutes < 0 {
		minutes = minutesInDay + minutes
	}
	return Clock{minutes}
}

func (clock Clock) String() string {
	return fmt.Sprintf("%02d:%02d", clock.Hour(), clock.Minute())
}

func (clock Clock) Add(minutes int) Clock {
	return New(0, clock.minute_of_the_day + minutes)
}

func (clock Clock) Hour() int {
	return clock.minute_of_the_day / 60
}

func (clock Clock) Minute() int {
	return clock.minute_of_the_day % 60
}
